### PR TITLE
feat: package component hooks for core24

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox run -m lint`?
- [x] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

Add support for packaging component hooks for core24.

I recommend reviewing per-commit.

(CRAFT-2540)

### CODE REVIEW

File: snapcraft/parts/setup_assets.py [hunk 1]
```diff
@@ -37,46 +37,38 @@ def setup_assets(
...
-        for hook in hooks_dir.iterdir():  # type: ignore
-            # mypy says we are passing a Hook, but this is indeed a Path
-            _ensure_hook_executable(hook)  # type: ignore
+    copy_assets(assets_dir, prime_dir, meta_directory_handler)
+    setup_hooks(project.hooks, prime_dir)
+
+    # core22 doesn't provide a meta_directory_handler, which means that core22 component
+    # hooks would be handled like core22 snap hooks (with hook wrappers).
+    # This behavior is not desired, so component hooks are currently ignored for core22.
+    if project.components and project.get_effective_base() != "core22":
```
Comment (mr-cal) [inline snapcraft/parts/setup_assets.py:64]: This and an updated core22 spread test will give us core22 component hook support.

```suggestion
    if project.components:
```

File: snapcraft/services/lifecycle.py [hunk 2]
```diff
@@ -97,6 +97,23 @@ def get_prime_dir(self, component: str | None = None) -> Path:
                 "because it does not exist."
             ) from err
 
+    @property
+    def prime_dirs(self) -> dict[str | None, Path]:
```
Comment (mr-cal) [inline snapcraft/services/lifecycle.py:101]: This is duplicated for core22 (`snapcraft/parts/parts.py`) and core24 (`snapcraft/services/lifecycle.py`).  There are a handful of duplicated functions between these two locations, so I decided to follow the same paradigm.

File: snapcraft/parts/parts.py [hunk 3]
```diff
@@ -143,6 +143,23 @@ def prime_dir(self) -> pathlib.Path:
         """Return the parts prime directory path."""
         return self._lcm.project_info.prime_dir
 
+    @property
+    def prime_dirs(self) -> dict[str | None, pathlib.Path]:
```
Comment (sergiusens) [inline snapcraft/parts/parts.py:147]: this should probably replace def get_prime_dir_for_component
this is also needs to be renames as it is not all prime_dirs, it is missing the default, so component_prime_dirs would be more accurate

File: snapcraft/parts/setup_assets.py [hunk 4]
```diff
@@ -119,6 +98,51 @@ def setup_assets(
             desktop_file.write(gui_dir=gui_dir, icon_path=relative_icon_path)
 
 
+def copy_assets(
+    assets_dir: Path,
+    prime_dir: Path,
+    meta_directory_handler: Optional[Callable[[Path, Path], None]] = None,
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply snapcraft/parts/setup_assets.py]: Yes, I updated the whole module to be consistent.

File: snapcraft/services/lifecycle.py (hunk 2 above)
Comment (mr-cal) [reply snapcraft/services/lifecycle.py:96]: I de-duplicated it.

**Review by tigarmo — APPROVED** [review]

File: tests/spread/core24/components/squashfs-root/meta/component.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+component: hello-components+share
```
Comment (sergiusens) [inline tests/spread/core24/components/squashfs-root/meta/component.yaml:1]: is this file and everything in squashfs-root necessary?

**Pull request merged**